hydra-node/src/Hydra/HeadLogic.hs

```diff
@@ -1041,39 +1071,43 @@ isLeader HeadParameters{parties} p (UnsafeSnapshotNumber sn) =
     _ -> False
 
 -- | Snapshot emission decider
-newSn :: IsTx tx => Environment -> HeadParameters -> CoordinatedHeadState tx -> SnapshotOutcome tx
+newSn :: Eq (SeenSnapshot tx) => Environment -> HeadParameters -> CoordinatedHeadState tx -> SnapshotOutcome tx
 newSn Environment{party} parameters CoordinatedHeadState{confirmedSnapshot, seenSnapshot, seenTxs} =
-  let Snapshot{number} = getSnapshot confirmedSnapshot
-      nextSnapshotNumber = succ number
-   in if
-          | not (isLeader parameters party nextSnapshotNumber) ->
-            ShouldNotSnapshot $ NotLeader nextSnapshotNumber
-          | seenSnapshot /= NoSeenSnapshot ->
-            ShouldNotSnapshot $ SnapshotInFlight nextSnapshotNumber
-          | null seenTxs ->
-            ShouldNotSnapshot NoTransactionsToSnapshot
-          | otherwise ->
-            ShouldSnapshot nextSnapshotNumber seenTxs
-
--- TODO: This is the only logic NOT in 'update' and gets applied on top of it in
--- "Hydra.Node". We tried to do this decision inside 'update' in the past, but
--- ended up doing it here. Is it really not possible to just call this function
--- from the respective places in 'update'? i.e. as a last step on
--- 'onOpenNetworkReqTx' and 'onOpenNetworkAckSn'?
-emitSnapshot :: IsTx tx => Environment -> [Effect tx] -> HeadState tx -> (HeadState tx, [Effect tx])
-emitSnapshot env@Environment{party} effects = \case
-  st@(Open OpenState{parameters, coordinatedHeadState, previousRecoverableState, chainState, headId}) ->
-    case newSn env parameters coordinatedHeadState of
-      ShouldSnapshot sn txs ->
-        ( Open
-            OpenState
-              { parameters
-              , coordinatedHeadState = coordinatedHeadState{seenSnapshot = RequestedSnapshot}
-              , previousRecoverableState
-              , chainState
-              , headId
-              }
-        , NetworkEffect (ReqSn party sn txs) : effects
-        )
-      _ -> (st, effects)
-  st -> (st, effects)
+  if
+      | not (isLeader parameters party nextSn) ->
+        ShouldNotSnapshot $ NotLeader nextSn
+      | -- REVIEW: This is slightly different than in the spec. Also, if we use
```

Comment:
Should we have a session to figure this out before merging?

Reply:
I already did (will be in a follow-up PR)